### ISSUE TITLE
dd-agent container stops collecting logs after some time

### ISSUE THREAD

**Issue by ibox4real** [DataDog/datadog-agent#12101]
**Output of the info page (if this is a bug)**
```
Getting the status from the agent.

===============
Agent (v7.34.0)
===============

  Status date: 2022-05-18 18:35:54.752 UTC (1652898954752)
  Agent start: 2022-05-17 14:29:06.232 UTC (1652797746232)
  Pid: 384
  Go Version: go1.16.12
  Python Version: 3.8.11
  Build arch: amd64
  Agent flavor: agent
  Check Runners: 4
  Log Level: info

  Paths
  =====
    Config File: /etc/datadog-agent/datadog.yaml
    conf.d: /etc/datadog-agent/conf.d
    checks.d: /etc/datadog-agent/checks.d

  Clocks
  ======
    NTP offset: -199µs
    System time: 2022-05-18 18:35:54.752 UTC (1652898954752)

  Host Info
  =========
    bootTime: 2022-03-22 10:08:02 UTC (1647943682000)
    kernelArch: x86_64
    kernelVersion: 4.14.256-197.484.amzn2.x86_64
    os: linux
    platform: ubuntu
    platformFamily: debian
    platformVersion: 21.10
    procs: 150
    uptime: 1348h21m11s
    virtualizationRole: guest
    virtualizationSystem: docker

  Hostnames
  =========
    ec2-hostname: ip-172-16-12-117.ec2.internal
    hostname: i-0d81ab1e31345325d
    instance-id: i-0d81ab1e31345325d
    socket-fqdn: 98b05023a476
    socket-hostname: 98b05023a476
    host tags:
      "env:production"
    hostname provider: aws
    unused hostname providers:
      azure: azure_hostname_style is set to 'os'
      configuration/environment: hostname is empty
      gce: unable to retrieve hostname from GCE: GCE metadata API error: status code 404 trying to GET http://169.254.169.254/computeMetadata/v1/instance/hostname

  Metadata
  ========
    agent_version: 7.34.0
    cloud_provider: AWS
    config_apm_dd_url:
    config_dd_url:
    config_logs_dd_url:
    config_logs_socks5_proxy_address:
    config_no_proxy: []
    config_process_dd_url:
    config_proxy_http:
    config_proxy_https:
    config_site:
    feature_apm_enabled: true
    feature_cspm_enabled: false
    feature_cws_enabled: false
    feature_logs_enabled: true
    feature_networks_enabled: false
    feature_process_enabled: true
    flavor: agent
    hostname_source: aws
    install_method_installer_version: docker
    install_method_tool: docker
    install_method_tool_version: docker
    logs_transport: HTTP

=========
Collector
=========

  Running Checks
  ==============

    container
    ---------
      Instance ID: container [OK]
      Configuration Source: file:/etc/datadog-agent/conf.d/container.d/conf.yaml.default
      Total Runs: 6,747
      Metric Samples: Last Run: 110, Total: 742,152
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 0, Total: 0
      Average Execution Time : 8ms
      Last Execution Date : 2022-05-18 18:35:45 UTC (1652898945000)
      Last Successful Execution Date : 2022-05-18 18:35:45 UTC (1652898945000)


    cpu
    ---
      Instance ID: cpu [OK]
      Configuration Source: file:/etc/datadog-agent/conf.d/cpu.d/conf.yaml.default
      Total Runs: 6,747
      Metric Samples: Last Run: 9, Total: 60,716
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 0, Total: 0
      Average Execution Time : 0s
      Last Execution Date : 2022-05-18 18:35:52 UTC (1652898952000)
      Last Successful Execution Date : 2022-05-18 18:35:52 UTC (1652898952000)


    disk (4.5.2)
    ------------
      Instance ID: disk:e5dffb8bef24336f [OK]
      Configuration Source: file:/etc/datadog-agent/conf.d/disk.d/conf.yaml.default
      Total Runs: 6,746
      Metric Samples: Last Run: 172, Total: 1,160,312
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 0, Total: 0
      Average Execution Time : 48ms
      Last Execution Date : 2022-05-18 18:35:44 UTC (1652898944000)
      Last Successful Execution Date : 2022-05-18 18:35:44 UTC (1652898944000)


    docker
    ------
      Instance ID: docker [OK]
      Configuration Source: file:/etc/datadog-agent/conf.d/docker.d/conf.yaml.default
      Total Runs: 6,747
      Metric Samples: Last Run: 166, Total: 1,120,006
      Events: Last Run: 1, Total: 9
      Service Checks: Last Run: 1, Total: 6,747
      Average Execution Time : 163ms
      Last Execution Date : 2022-05-18 18:35:51 UTC (1652898951000)
      Last Successful Execution Date : 2022-05-18 18:35:51 UTC (1652898951000)


    file_handle
    -----------
      Instance ID: file_handle [OK]
      Configuration Source: file:/etc/datadog-agent/conf.d/file_handle.d/conf.yaml.default
      Total Runs: 6,746
      Metric Samples: Last Run: 5, Total: 33,730
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 0, Total: 0
      Average Execution Time : 0s
      Last Execution Date : 2022-05-18 18:35:43 UTC (1652898943000)
      Last Successful Execution Date : 2022-05-18 18:35:43 UTC (1652898943000)


    io
    --
      Instance ID: io [OK]
      Configuration Source: file:/etc/datadog-agent/conf.d/io.d/conf.yaml.default
      Total Runs: 6,747
      Metric Samples: Last Run: 41, Total: 276,600
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 0, Total: 0
      Average Execution Time : 0s
      Last Execution Date : 2022-05-18 18:35:50 UTC (1652898950000)
      Last Successful Execution Date : 2022-05-18 18:35:50 UTC (1652898950000)


    load
    ----
      Instance ID: load [OK]
      Configuration Source: file:/etc/datadog-agent/conf.d/load.d/conf.yaml.default
      Total Runs: 6,746
      Metric Samples: Last Run: 6, Total: 40,476
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 0, Total: 0
      Average Execution Time : 0s
      Last Execution Date : 2022-05-18 18:35:42 UTC (1652898942000)
      Last Successful Execution Date : 2022-05-18 18:35:42 UTC (1652898942000)


    memory
    ------
      Instance ID: memory [OK]
      Configuration Source: file:/etc/datadog-agent/conf.d/memory.d/conf.yaml.default
      Total Runs: 6,747
      Metric Samples: Last Run: 20, Total: 134,940
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 0, Total: 0
      Average Execution Time : 0s
      Last Execution Date : 2022-05-18 18:35:49 UTC (1652898949000)
      Last Successful Execution Date : 2022-05-18 18:35:49 UTC (1652898949000)


    network (2.5.0)
    ---------------
      Instance ID: network:d884b5186b651429 [OK]
      Configuration Source: file:/etc/datadog-agent/conf.d/network.d/conf.yaml.default
      Total Runs: 6,746
      Metric Samples: Last Run: 129, Total: 870,226
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 0, Total: 0
      Average Execution Time : 8ms
      Last Execution Date : 2022-05-18 18:35:41 UTC (1652898941000)
      Last Successful Execution Date : 2022-05-18 18:35:41 UTC (1652898941000)


    ntp
    ---
      Instance ID: ntp:d884b5186b651429 [OK]
      Configuration Source: file:/etc/datadog-agent/conf.d/ntp.d/conf.yaml.default
      Total Runs: 113
      Metric Samples: Last Run: 1, Total: 113
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 1, Total: 113
      Average Execution Time : 1ms
      Last Execution Date : 2022-05-18 18:29:15 UTC (1652898555000)
      Last Successful Execution Date : 2022-05-18 18:29:15 UTC (1652898555000)


    uptime
    ------
      Instance ID: uptime [OK]
      Configuration Source: file:/etc/datadog-agent/conf.d/uptime.d/conf.yaml.default
      Total Runs: 6,747
      Metric Samples: Last Run: 1, Total: 6,747
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 0, Total: 0
      Average Execution Time : 0s
      Last Execution Date : 2022-05-18 18:35:48 UTC (1652898948000)
      Last Successful Execution Date : 2022-05-18 18:35:48 UTC (1652898948000)

========
JMXFetch
========

  Information
  ==================
  Initialized checks
  ==================
    no checks

  Failed checks
  =============
    no checks

=========
Forwarder
=========

  Transactions
  ============
    Cluster: 0
    ClusterRole: 0
    ClusterRoleBinding: 0
    CronJob: 0
    DaemonSet: 0
    Deployment: 0
    Dropped: 0
    HighPriorityQueueFull: 0
    Job: 0
    Node: 0
    PersistentVolume: 0
    PersistentVolumeClaim: 0
    Pod: 0
    ReplicaSet: 0
    Requeued: 1
    Retried: 1
    RetryQueueSize: 0
    Role: 0
    RoleBinding: 0
    Service: 0
    ServiceAccount: 0
    StatefulSet: 0

  Transaction Successes
  =====================
    Total number: 14233
    Successes By Endpoint:
      check_run_v1: 6,747
      intake: 570
      metadata_v1: 169
      series_v1: 6,747

  Transaction Errors
  ==================
    Total number: 1
    Errors By Type:

  HTTP Errors
  ==================
    Total number: 1
    HTTP Errors By Code:
      504: 1

  On-disk storage
  ===============
    On-disk storage is disabled. Configure `forwarder_storage_max_size_in_bytes` to enable it.

  API Keys status
  ===============
    API key ending with 8dca3: API Key valid

==========
Endpoints
==========
  https://app.datadoghq.com - API Key ending with:
      - 8dca3

==========
Logs Agent
==========

    Reliable: Sending compressed logs in HTTPS to agent-http-intake.logs.datadoghq.com on port 443
    BytesSent: 2.556281421e+09
    EncodedBytesSent: 1.09956895e+08
    LogsProcessed: 1.105035e+06
    LogsSent: 1.105035e+06

  docker
  ------
    - Type: docker
      Status: OK
      Inputs:
        f154b44857ae96d142cd6e8ba631cf498f71a2adb1d0b7aa9b7c3f3fd1b65fa3
      BytesRead: 1.15379e+06
      Average Latency (ms): 0
      24h Average Latency (ms): 0
      Peak Latency (ms): 11
      24h Peak Latency (ms): 11
    - Type: docker
      Status: Pending
      BytesRead: 0
      Average Latency (ms): 0
      24h Average Latency (ms): 0
      Peak Latency (ms): 0
      24h Peak Latency (ms): 0

=========
APM Agent
=========
  Status: Running
  Pid: 379
  Uptime: 101207 seconds
  Mem alloc: 8,866,384 bytes
  Hostname: i-0d81ab1e31345325d
  Receiver: 0.0.0.0:8126
  Endpoints:
    https://trace.agent.datadoghq.com

  Receiver (previous minute)
  ==========================
    No traces received in the previous minute.
    Default priority sampling rate: 100.0%

  Writer (previous minute)
  ========================
    Traces: 0 payloads, 0 traces, 0 events, 0 bytes
    Stats: 0 payloads, 0 stats buckets, 0 bytes

=========
Aggregator
=========
  Checks Metric Sample: 4,581,176
  Dogstatsd Metric Sample: 826,487
  Event: 10
  Events Flushed: 10
  Number Of Flushes: 6,747
  Series Flushed: 4,041,040
  Service Check: 74,439
  Service Checks Flushed: 81,186
=========
DogStatsD
=========
  Event Packets: 0
  Event Parse Errors: 0
  Metric Packets: 826,486
  Metric Parse Errors: 0
  Service Check Packets: 0
  Service Check Parse Errors: 0
  Udp Bytes: 62,313,511
  Udp Packet Reading Errors: 0
  Udp Packets: 590,733
  Uds Bytes: 0
  Uds Origin Detection Errors: 0
  Uds Packet Reading Errors: 0
  Uds Packets: 0
  Unterminated Metric Errors: 0

=============
Autodiscovery
=============
  Enabled Features
  ================
    docker

```

**Describe what happened:**
Log collection on one of the containers stops after some time. The container is running as normal and the logs are going to stdout but the agent is not collecting them anymore. Status page shows Pending for some reason. 

**Additional environment details (Operating System, Cloud provider, etc):**
 Amazon Linux 2 (4.14.256)
 Docker 20.10.7
 
 ```
  dd-agent:
    image: datadog/agent
    environment:
      DD_APM_ENABLED: 'true'
      DD_APM_NON_LOCAL_TRAFFIC: 'true'
      DD_LOGS_ENABLED: 'true'
      DD_PROCESS_AGENT_ENABLED: 'true'
      DD_TAGS: '"env:production"'
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock:ro
      - /proc/:/host/proc/:ro
      - /sys/fs/cgroup/:/host/sys/fs/cgroup:ro
      - /opt/datadog-agent/run:/opt/datadog-agent/run:rw
 ```

The log collection seems to get fixed whenever i restart the target container. Not sure what causes it, but it keeps happening agfter a while.

**Comment by carlosroman** [comment]
We've made some major improvements in the way the Agent tails containers in general. This should improve the way we pick up logs and hopefully stop the Agent ending up in a `pending` state.

These changes will be released in next version of the Agent (`7.40`). You can try out a preview of these changes in `7.39` by enabling a feature flag.

```
environment:
  DD_LOGS_CONFIG_CCA_IN_AD:  'true'
```

**Comment by maokitty** [comment]
there is another case, some agent which is Pending  still sending data to backend, while others is stuck,it's so weired

**Comment by cfernhout** [comment]
Got this issue quite often after upgrading to 7.40.1. The trigger is (re)deploying new containers.

**Comment by cfernhout** [comment]
@carlosroman are there any improvements concerning this issue in Agent `7.41`?

**Comment by grickit** [comment]
I began seeing this issue after a datadog-agent upgrade in December.

When restarting containers Datadog would stop reporting certain stats and stop collecting logs. We rolled back to 7.38.2 from August to see if that helps and created the following monitor that tracks the issue in the meantime:

`avg(last_5m):sum:docker.containers.running{!image_name:*} by {host} > 1`

`{{#is_alert}}At least one docker container on {{host.name}} is not reporting its image_name. This likely means that datadog-agent cannot see it.{{/is_alert}}`

**Comment by grickit** [comment]
This issue seems to be a duplicate: https://github.com/DataDog/datadog-agent/issues/8184 (it came first, but has less details)

**Comment by grickit** [comment]
I should have come back sooner to confirm for yall that 7.38.2 does not seem to be affected by this issue. You can downgrade by rerunning the install script with these environment variables:

`DD_AGENT_MAJOR_VERSION=7 DD_AGENT_MINOR_VERSION=38`

**Comment by ilibar-zpt** [comment]
Experiencing similar problem, suspecting a bug with logs collection when running as host agent and using docker socket.
Moved agent into container and mounted all docker things so it collects using file, seems fine so far

**Comment by carlosroman** [comment]
We did have a bug with our Docker socket tailer (#15138) which has been fixed and released in `7.43.0`. I'd be interested to know if this solves people's issues or not.

**Comment by joedenniss** [comment]
We were experiencing this even on the latest Datadog agent version. The crucial insight for us was that the Datadog agent is tailing the logs with the `since` flag, and we observed that if we ran `docker logs` with the `--since` flag against the containers in question it would hang indefinitely, similar to what is described here: https://github.com/moby/moby/issues/38640#issuecomment-500425221

TL;DR our container logs were GBs in size and configuring log rotation solved the problem for us.

**Comment by grickit** [comment]
As far as I know we haven't seen this issue since upgrading to agent 7.43.2 many months ago. Sorry for the delay on reporting again.
